nestkernel/eprop_archiving_node.cpp

```diff
@@ -0,0 +1,300 @@
+    }
+  }
+}
+
+
+void
+nest::EpropArchivingNode::get_spike_history( double t1,
+  double t2,
+  std::deque< double >::iterator* start,
+  std::deque< double >::iterator* finish )
+{
+  // set pointer to spike history entries corresponding to times t1 and t2
+  *finish = spike_history_.end();
+
+  if ( spike_history_.empty() )
+  {
+    *start = *finish;
+    return;
+  }
+  else
+  {
+    std::deque< double >::iterator runner1 =
+      std::lower_bound( spike_history_.begin(), spike_history_.end(), t1 + eps_ );
+
+    *start = runner1;
+
+    std::deque< double >::iterator runner2 = std::lower_bound( runner1, spike_history_.end(), t2 + eps_ );
+
+    *finish = runner2;
+  }
+}
+
+void
+nest::EpropArchivingNode::tidy_eprop_history()
+{
+  double earliest_time_to_keep = ( t_last_update_per_synapse_.begin() )->t_;
+  if ( !eprop_history_.empty() )
+  {
+    std::deque< histentry_eprop >::iterator start;
+    std::deque< histentry_eprop >::iterator finish;
+
+    // find eprop history entries for times earlier than earliest last update time
+    nest::EpropArchivingNode::get_eprop_history( 0.0, earliest_time_to_keep, &start, &finish );
+
+    eprop_history_.erase( eprop_history_.begin(), finish ); // erase found entries since no longer used
+  }
+}
+
+void
+nest::EpropArchivingNode::tidy_spike_history()
+{
+  double earliest_time_to_keep = ( t_last_update_per_synapse_.begin() )->t_;
+  while ( ( !spike_history_.empty() ) && ( spike_history_.front() + 1.0e-6 < earliest_time_to_keep ) )
+  {
+    spike_history_.pop_front();
+  }
+}
+
+void
+nest::EpropArchivingNode::write_v_m_pseudo_deriv_to_eprop_history( long time_step, double v_m_pseudo_deriv )
+{
+  if ( n_incoming_ )
+  {
+    const Time time_step_ = Time::step( time_step );
+    const double time_ms = time_step_.get_ms();
+    eprop_history_.push_back( histentry_eprop( time_ms, v_m_pseudo_deriv, 0.0 ) );
+  }
+}
+
+void
+nest::EpropArchivingNode::write_error_signal_to_eprop_history( long time_step, double error_signal )
+{
+  const Time time_step_ = Time::step( time_step );
+  const double time_ms = time_step_.get_ms();
+  eprop_history_.push_back( histentry_eprop( time_ms, 0.0, error_signal ) );
+}
+
+void
+nest::EpropArchivingNode::write_spike_history( long time_step )
+{
+  const Time time_step_ = Time::step( time_step );
+  const double time_ms = time_step_.get_ms();
+  spike_history_.push_back( time_ms );
+}
+
+double
+nest::EpropArchivingNode::get_learning_signal_from_eprop_history( long unsigned int shift )
+{
+  if ( eprop_history_.size() > shift )
+  {
+    return ( ( eprop_history_.rbegin() ) + shift )->learning_signal_;
+  }
+  else
+  {
+    return 0.0;
+  }
+}
+
+void
+nest::EpropArchivingNode::write_learning_signal_to_eprop_history( LearningSignalConnectionEvent& e )
+{
+  const double weight = e.get_weight();
+  const long delay = e.get_delay_steps();
+  const Time stamp = e.get_stamp();
+
+  double t_ms = stamp.get_ms() - 2.0 * Time::get_resolution().get_ms();
```

Comment:
Why hard-coded factor 2?

Reply:
Fixed in https://github.com/nest/nest-simulator/pull/2867/commits/02150742b00cf1030d1f079e684b1d0af5f1dbdf